irods/files/rule-bases/ipc-custom.re

```diff
@@ -621,3 +621,147 @@ pep_resource_resolve_hierarchy_pre(*INSTANCE, *CONTEXT, *OUT, *OPERATION, *HOST,
+  }
+}
+
+pep_api_rm_coll_pre(*INSTANCE, *COMM, *RMCOLLINP, *COLLOPRSTAT) {
+  if (errorcode(*RMCOLLINP.forceFlag) != 0) {
+    msiGetSystemTime(*timestamp, "");
+    *collNamePath = *RMCOLLINP.coll_name;
+    *timestampVar = _ipc_mkDataObjAndCollTimestampVar(/*collNamePath);
+    temporaryStorage.'*timestampVar' = *timestamp;
+    msiModAVUMetadata("-C", *collNamePath, "set", "ipc::trash_timestamp", *timestamp, "");
+  }
+}
+
+pep_api_rm_coll_except(*INSTANCE, *COMM, *RMCOLLINP, *COLLOPRSTAT) {
+  *collNamePath = *RMCOLLINP.coll_name;
+  *timestampVar = _ipc_mkDataObjAndCollTimestampVar(/*collNamePath);
+  if (errorcode(temporaryStorage.'*timestampVar') == 0) {
+    msiModAVUMetadata("-C", *collNamePath, "rm", "ipc::trash_timestamp", temporaryStorage.'*timestampVar', "");
+  }
+}
+
+pep_api_coll_create_post(*INSTANCE, *COMM, *COLLCREATEINP) {
+  *zone = ipc_ZONE;
+  if (*COLLCREATEINP.coll_name like '/*zone/trash/*') {
+    msiGetSystemTime(*timestamp, "");
+    msiModAVUMetadata("-C", *COLLCREATEINP.coll_name, "set", "ipc::trash_timestamp", *timestamp, "");
+  }
+}
+
+pep_api_data_obj_rename_pre(*INSTANCE, *COMM, *DATAOBJRENAMEINP) {
+  *zone = ipc_ZONE;
+  if ((*DATAOBJRENAMEINP.src_obj_path like '/*zone/trash/*') && (*DATAOBJRENAMEINP.dst_obj_path not like '/*zone/trash/*')) {
+    *srcObjPath = *DATAOBJRENAMEINP.src_obj_path;
+    *timestampVar = _ipc_mkDataObjAndCollTimestampVar(/*srcObjPath);
+    msiGetObjType(*srcObjPath, *Type);
+    if (ipc_isCollection(*Type)) {
+      foreach(*Row in SELECT META_COLL_ATTR_VALUE
+                        WHERE COLL_NAME like '*srcObjPath'
+                          AND META_COLL_ATTR_NAME = 'ipc::trash_timestamp') {
+                            *row_meta_coll_attr_value = *Row.META_COLL_ATTR_VALUE;
+                            temporaryStorage.'*timestampVar' = *row_meta_coll_attr_value;
+      }
+    }
+    else if (ipc_isDataObject(*Type)) {
+      msiSplitPath(*srcObjPath, *Coll, *File);
+      foreach(*Row in SELECT META_DATA_ATTR_VALUE
+                        WHERE COLL_NAME like '*Coll'
+                          AND DATA_NAME like '*File'
+                            AND META_DATA_ATTR_NAME = 'ipc::trash_timestamp') {
+                              *row_meta_data_attr_value = *Row.META_DATA_ATTR_VALUE;
+                              temporaryStorage.'*timestampVar' = *row_meta_data_attr_value;
+      }
+    }
+  }
+}
+
+pep_api_data_obj_rename_post(*INSTANCE, *COMM, *DATAOBJRENAMEINP) {
+  *zone = ipc_ZONE;
+  if (*DATAOBJRENAMEINP.dst_obj_path like '/*zone/trash/*') {
+    msiGetSystemTime(*timestamp, "");
+    msiGetObjType(*DATAOBJRENAMEINP.dst_obj_path, *Type);
+    if (ipc_isDataObject(*Type)) {
+      msiModAVUMetadata("-d", *DATAOBJRENAMEINP.dst_obj_path, "set", "ipc::trash_timestamp", *timestamp, "");
+    }
+    else if (ipc_isCollection(*Type)) {
+      msiModAVUMetadata("-C", *DATAOBJRENAMEINP.dst_obj_path, "set", "ipc::trash_timestamp", *timestamp, "");
+    }
```

Comment:
Since the first argument to `msiModAVUMetadata` is the type of entity whose AVU is being modified, you could pass `*Type` in to both calls to this microservice.  If you do this, you'll see that they have the same form and you don't need the if-else branching.  By also using our function `ipc_getEntityType` from `ipc-services.re`, all of this reduces to 

```
msiModAVUMetadata(
  ipc_getEntityType(*DATAOBJRENAMEINP.dst_obj_path), 
  *DATAOBJRENAMEINP.dst_obj_path, 
  "set", 
  "ipc::trash_timestamp", 
  *timestamp, 
  "" );
```

Reply:
I was able to reduce ~10 LOC, thank you.